### PR TITLE
Add EvalOnDartLibrary utilities

### PR DESCRIPTION
**Pull request by rrousselGit**
The `Disposable` interface separates `EvalOnDartLibrary` from `ObjectGroup` which is specific to the inspector service.

`asyncEval` allows optionally using `await` inside an expression, in which case the expression will not resolve until the future awaited completes.

### CODE REVIEW

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 1]
```diff
@@ -1,11 +1,13 @@
-// Copyright 2018 The Chromium Authors. All rights reserved.
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:1]: leave existing copyright dates unchanged. only new files get new copyright dates.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 2]
```diff
@@ -156,25 +178,229 @@ class EvalOnDartLibrary {
...
+          scope: scope,
+          exception: result,
+        );
+      }
+    } catch (err, stack) {
+      _handleError(err, stack);
+      rethrow;
+    }
+
+    // We throw if the request was canceled instead of returning `null` because
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:391]: are you sure this is the case? You should get an InstanceRef with a null value not null. I think you might have been seeing null getting returned when eval was cancelled.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 3]
```diff
@@ -14,6 +16,22 @@ import 'globals.dart';
 import 'inspector/inspector_service.dart';
 import 'vm_service_wrapper.dart';
 
+abstract class Disposable {
+  bool get disposed;
+
+  void dispose();
+}
+
+class IsAlive extends Disposable {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:25]: I'm not seeing a need for two classes in this CL. Just make this the default implementation of Disposable and annotate dispose with mustCallSuper(). In dart someone can implement a class if they need to use it as an interface.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 4]
```diff
@@ -95,28 +114,31 @@ class EvalOnDartLibrary {
...
+    @required Disposable isAlive,
     Map<String, String> scope,
   }) {
     return addRequest(isAlive, () => _eval(expression, scope: scope));
   }
 
+  Future<LibraryRef> _waitForLibraryRef() async {
+    while (true) {
+      final libraryRef = await _libraryRef.future;
+      if (_libraryRef.isCompleted) {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:126]: this is a good race condition to check for but you should check for it by checking whether the isolateRef matches the isolateRef when the method started.

Comment (rrousselGit) [reply packages/devtools_app/lib/src/eval_on_dart_library.dart:126]: I didn't write this code. It was an "extract as method" from `eval`

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 5]
```diff
@@ -156,25 +178,229 @@ class EvalOnDartLibrary {
...
+  }) async {
+    final ref = await safeEval(expression, isAlive: isAlive, scope: scope);
+    if (ref == null) return null;
+
+    return getInstance(ref, isAlive);
+  }
+
+  static int _nextAsyncEvalId = 0;
+
+  EvalOnDartLibrary _dartDeveloperEvalCache;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:227]: you should dispose these caches if they are non null in the dispose method for this eval.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 6]
```diff
@@ -156,25 +178,229 @@ class EvalOnDartLibrary {
...
+
+        return await service.evaluate(
+          isolateId,
+          libraryRef.id,
+          expression,
+          scope: scope,
+        );
+      });
+
+      if (result == null) return null;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/eval_on_dart_library.dart:363]: this case is where the EvalOnDartLibrary was disposed while the request was in progress. Do you want to throw an error here or only if it was disposed before the request started?

**Review by jacob314 — COMMENTED** [review]
Thanks for adding all this great functionality to eval!
A few minor nits.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart (hunk 2 above)
Comment (rrousselGit) [reply packages/devtools_app/lib/src/eval_on_dart_library.dart:391]: Upon testing it, it appears that `eval('null')` does return an InstanceRef.

But `eval('[null]')` followed by a `getInstance` and reading `instance.elements.first` returns `null` (and I believe the issue is the same for maps)

So it appears to be inconsistent

**Comment by jacob314** [comment]
lgtm

File: packages/devtools_app/lib/src/eval_on_dart_library.dart (hunk 4 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/eval_on_dart_library.dart:121]: feel free to just make this a TODO for me to fix.

File: packages/devtools_app/lib/src/eval_on_dart_library.dart [hunk 7]
```diff
@@ -156,25 +178,229 @@ class EvalOnDartLibrary {
...
+    @required Disposable isAlive,
+    Map<String, String> scope,
+  }) async {
+    final id = _nextAsyncEvalId++;
+
+    // start awaiting the event before starting the evaluation, in case the
+    // event is received before the eval function completes.
+    final future = serviceManager.service.onExtensionEvent.firstWhere((event) {
+      return event.extensionKind == 'future_completed' &&
+          event.extensionData.data['id'] == id;
```
[the remark this answers is not in the document]
Comment (rrousselGit) [reply packages/devtools_app/lib/src/eval_on_dart_library.dart]: Done

File: packages/devtools_app/lib/src/eval_on_dart_library.dart (hunk 2 above)
Comment (rrousselGit) [reply packages/devtools_app/lib/src/eval_on_dart_library.dart]: I've removed the null checks and updated the comment

**Comment by rrousselGit** [comment]
I have fixed the different nits so this should be good to go I believe

File: packages/devtools_app/lib/src/eval_on_dart_library.dart (hunk 2 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/eval_on_dart_library.dart]: Sounds good. Lets file a bug to get the VM Service consistent on this point. I think you have discovered a bug in the VM Service api or a micro-optimization that makes the api less consistent.

**Review by jacob314 — APPROVED** [review]
lgtm

**Pull request merged**